### PR TITLE
Msf::Post::Windows::Accounts: Add methods to retrieve domain / domain controller

### PR DESCRIPTION
**Pull request by bcoles**
I will PR a few improvements to existing modules which will utilise these methods.

### CODE REVIEW

File: lib/msf/core/post/windows/accounts.rb [hunk 1]
```diff
@@ -66,11 +66,40 @@ def initialize(info = {})
...
         end
 
-        # Check if target is a domain controller
+        # Check if host is an Active Directory domain controller
         #
-        # @return [Boolean] Target host is a domain controller
+        # @return [Boolean] Target host is an Active Directory domain controller
         def domain_controller?
-          registry_key_exist?('HKLM\\SYSTEM\\CurrentControlSet\\Services\\NTDS\\Parameters')
+          registry_enumkeys("HKLM\\SYSTEM\\CurrentControlSet\\Services\\NTDS")&.include?('Parameters') ? true : false
```
[the remark this answers is not in the document]
Comment (bcoles) [reply lib/msf/core/post/windows/accounts.rb:73]: Yes. https://github.com/rapid7/metasploit-framework/pull/16898#discussion_r956727498

[the remark this answers is not in the document]
Comment (cdelafuente-r7) [reply lib/msf/core/post/windows/accounts.rb:73]: Oh, sorry, I missed this comment. It makes sense.

**Comment by bcoles** [comment]
> `get_domain_name` returns `nil` on my shell session running on a Windows 2003x64 DC
> It does that because there is no `MachineDomain` key in the registry for 2003x64

Unfortunate. Confirmed failure on Windows Server 2003 (x86).

There are other techniques to retrieve the domain name. For example, the mess of approaches use by `post/windows/gather/credentials/gpp`:

https://github.com/rapid7/metasploit-framework/blob/3682f4dee9e4f1718c205a448638f16a9cff57ea/modules/post/windows/gather/credentials/gpp.rb#L376-L389

These registry keys (and several others) could be incorporated into the library. The techniques above identify domain names - not necessarily the primary domain.

I'd rather create an issue for this and move on. There are other PRs which depend on this PR and many other bugs and features which are more useful than supporting Windows 2003.

Prior to this PR, the domain name and domain controller can be retrieved only on Meterpreter sessions. The changes introduced in this PR do not break the Meterpreter support and expand support to include non-Meterpreter sessions. This approach should work on Windows Server since 2008 onward - the last 14 years of server releases.

**Comment by bwatters-r7** [comment]
Makes sense to me.  Most coverage for shells is better than none...

**Comment by bwatters-r7** [comment]
# Release Notes
This PR improves the `domain_controller?` method to allow lower-priv users to invoke it, extends it to support shell sessions, and adds additional useful domain controller enumeration methods to the library.

**Comment by bcoles** [comment]
> Makes sense to me. Most coverage for shells is better than none...

That's the motivation and intent. That, and abstracting duplicated code in to reusable methods.

I've created these methods to decrease copy pasta. The existing copy pasta has the same issue in that it fails on Windows Server 2003.

Sample copy pasta:

https://github.com/rapid7/metasploit-framework/blob/2e89006c1c260fefd9279149d38af8afe1ba6251/modules/post/windows/gather/enum_computers.rb#L70-L83

https://github.com/rapid7/metasploit-framework/blob/28eab4d8710161470c715027116134fb15fb9613/modules/post/windows/gather/enum_domain_tokens.rb#L66-L76

https://github.com/rapid7/metasploit-framework/blob/2e89006c1c260fefd9279149d38af8afe1ba6251/modules/post/windows/gather/enum_tokens.rb#L105-L122

**Pull request merged**